src/session_lock/mod.rs

```diff
@@ -0,0 +1,211 @@
+    locked: AtomicBool,
+}
+
+impl Drop for SessionLockInner {
+    fn drop(&mut self) {
+        if self.locked.load(Ordering::SeqCst) {
+            self.session_lock.unlock_and_destroy();
+        } else {
+            self.session_lock.destroy();
+        }
+    }
+}
+
+/// A session lock
+///
+/// The lock is destroyed on drop, which must be done after `locked` or `finished`
+/// is received.
+#[must_use]
+#[derive(Debug, Clone)]
+pub struct SessionLock(Arc<SessionLockInner>);
+
+impl SessionLock {
+    pub fn from_ext_session_lock(surface: &ext_session_lock_v1::ExtSessionLockV1) -> Option<Self> {
+        surface.data::<SessionLockData>().and_then(|data| data.inner.upgrade()).map(Self)
+    }
+
+    pub fn is_locked(&self) -> bool {
+        self.0.locked.load(Ordering::SeqCst)
+    }
+}
+
+#[derive(Debug)]
+pub struct SessionLockData {
+    inner: Weak<SessionLockInner>,
+}
+
+impl SessionLock {
+    pub fn create_lock_surface<D>(
+        &self,
+        surface: impl Into<Surface>,
+        output: &wl_output::WlOutput,
+        qh: &QueueHandle<D>,
+    ) -> SessionLockSurface
+    where
+        D: Dispatch<ext_session_lock_surface_v1::ExtSessionLockSurfaceV1, SessionLockSurfaceData>
+            + 'static,
+    {
+        // Freeze the queue during the creation of the Arc to avoid a race between events on the
+        // new objects being processed and the Weak in the PopupData becoming usable.
+        let freeze = qh.freeze();
+        let surface = surface.into();
+
+        let inner = Arc::new_cyclic(|weak| {
+            let session_lock_surface = self.0.session_lock.get_lock_surface(
+                surface.wl_surface(),
+                output,
+                qh,
+                SessionLockSurfaceData { inner: weak.clone() },
+            );
+
+            SessionLockSurfaceInner { surface, session_lock_surface }
+        });
+        drop(freeze);
+
+        SessionLockSurface(inner)
+    }
+}
+
+/// A handler for [`ext_session_lock_manager_v1::ExtSessionLockManagerV1`]
+#[derive(Debug)]
+pub struct SessionLockState {
+    session_lock_manager: GlobalProxy<ext_session_lock_manager_v1::ExtSessionLockManagerV1>,
+}
+
+impl SessionLockState {
+    pub fn new<D>(globals: &GlobalList, qh: &QueueHandle<D>) -> Self
+    where
+        D: Dispatch<ext_session_lock_manager_v1::ExtSessionLockManagerV1, GlobalData> + 'static,
+    {
+        let session_lock_manager = GlobalProxy::from(globals.bind(qh, 1..=1, GlobalData));
+        Self { session_lock_manager }
+    }
+
+    pub fn lock<D>(&self, qh: &QueueHandle<D>) -> Result<SessionLock, GlobalError>
+    where
+        D: Dispatch<ext_session_lock_v1::ExtSessionLockV1, SessionLockData> + 'static,
+    {
+        let session_lock_manager = self.session_lock_manager.get()?;
+
+        // Freeze the queue during the creation of the Arc to avoid a race between events on the
+        // new objects being processed and the Weak in the PopupData becoming usable.
```

Comment:
I guess this comment is supposed to say `SessionLockData`?

Reply:
Ah right.

Though I copied this from `src/shell/wlr_layer/mod.rs`, so it was already a copy-paste error there. And in `src/shell/xdg/mod.rs` as well, it looks like.

I guess there's not a clear way to abstract this logic rather than copying it...

Reply:
Updated the comment here, and in the other places.